drivers/net/nsos_adapt.c

```diff
@@ -0,0 +1,673 @@
+		      const struct nsos_mid_sockaddr *addr_mid, size_t addrlen_mid)
+{
+	struct sockaddr_storage addr_storage;
+	struct sockaddr *addr = (struct sockaddr *)&addr_storage;
+	socklen_t addrlen;
+	int ret;
+
+	ret = sockaddr_from_nsos_mid(&addr, &addrlen, addr_mid, addrlen_mid);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = sendto(fd, buf, len,
+		     socket_flags_from_nsos_mid(flags) | MSG_NOSIGNAL,
+		     addr, addrlen);
+	if (ret < 0) {
+		return -errno_to_nsos_mid(errno);
+	}
+
+	return ret;
+}
+
+int nsos_adapt_recvfrom(int fd, void *buf, size_t len, int flags,
+			struct nsos_mid_sockaddr *addr_mid, size_t *addrlen_mid)
+{
+	struct sockaddr_storage addr_storage;
+	struct sockaddr *addr = (struct sockaddr *)&addr_storage;
+	socklen_t addrlen = sizeof(addr_storage);
+	int ret;
+
+	ret = recvfrom(fd, buf, len, socket_flags_from_nsos_mid(flags),
+		       addr, &addrlen);
+	if (ret < 0) {
+		return -errno_to_nsos_mid(errno);
+	}
+
+	sockaddr_to_nsos_mid(addr, addrlen, addr_mid, addrlen_mid);
+
+	return ret;
+}
+
+static uint64_t nsos_adapt_poll_time = NSI_NEVER;
+
+void nsos_adapt_poll_add(struct nsos_mid_pollfd *pollfd)
+{
+	sys_dlist_append(&nsos_adapt_sockets, &pollfd->host_node);
+
+	nsos_adapt_poll_time = nsi_hws_get_time() + 1;
+	nsi_hws_find_next_event();
+}
+
+void nsos_adapt_poll_remove(struct nsos_mid_pollfd *pollfd)
+{
+	sys_dlist_remove(&pollfd->host_node);
+}
+
+struct nsos_addrinfo_wrap {
+	struct nsos_mid_addrinfo addrinfo_mid;
+	struct nsos_mid_sockaddr_storage addr_storage;
+	struct addrinfo *addrinfo;
+};
+
+static int addrinfo_to_nsos_mid(struct addrinfo *res,
+				struct nsos_mid_addrinfo **mid_res)
+{
+	struct nsos_addrinfo_wrap *nsos_res_wraps;
+	size_t idx_res = 0;
+	size_t n_res = 0;
+	int ret;
+
+	for (struct addrinfo *res_p = res; res_p; res_p = res_p->ai_next) {
+		n_res++;
+	}
+
+	if (n_res == 0) {
+		return 0;
+	}
+
+	nsos_res_wraps = calloc(n_res, sizeof(*nsos_res_wraps));
+	if (!nsos_res_wraps) {
+		return -NSOS_MID_ENOMEM;
+	}
+
+	for (struct addrinfo *res_p = res; res_p; res_p = res_p->ai_next, idx_res++) {
+		struct nsos_addrinfo_wrap *wrap = &nsos_res_wraps[idx_res];
+
+		wrap->addrinfo = res_p;
+
+		wrap->addrinfo_mid.ai_flags = res_p->ai_flags;
+
+		ret = socket_family_to_nsos_mid(res_p->ai_family, &wrap->addrinfo_mid.ai_family);
+		if (ret < 0) {
+			goto free_wraps;
+		}
+
+		ret = socket_type_to_nsos_mid(res_p->ai_socktype, &wrap->addrinfo_mid.ai_socktype);
+		if (ret < 0) {
+			goto free_wraps;
+		}
+
+		ret = socket_proto_to_nsos_mid(res_p->ai_protocol, &wrap->addrinfo_mid.ai_protocol);
+		if (ret < 0) {
+			goto free_wraps;
+		}
+
+		wrap->addrinfo_mid.ai_addr =
+			(struct nsos_mid_sockaddr *)&wrap->addr_storage;
+		wrap->addrinfo_mid.ai_addrlen = sizeof(wrap->addr_storage);
+
+		sockaddr_to_nsos_mid(res_p->ai_addr, res_p->ai_addrlen,
+				     wrap->addrinfo_mid.ai_addr, &wrap->addrinfo_mid.ai_addrlen);
+
+		wrap->addrinfo_mid.ai_canonname =
```

Comment:
maybe a stupid question, but: when will this be free'd?

Reply:
Not a stupid queston at all. This was indeed a memory leak. Added:
```
for (struct nsos_mid_addrinfo *res_p = res_mid; res_p; res_p = res_p->ai_next) {
	free(res_p->ai_canonname);
}
```
in `nsos_adapt_freeaddrinfo()`. Updated `nsos_sockets.c` as well.